`clippy::obfuscated_if_else` fix fails

### Summary

I ran clippy fix on my code and it had to revert.

https://github.com/diondokter/nrf-modem/blob/b88efe35f6abe19515e7ae3be8ac6431d55164c9/src/gnss.rs#L217-L239 

### Reproducer

I tried this code:

```rust
impl From<NmeaMask> for u16 {
    fn from(mask: NmeaMask) -> Self {
        mask.gga
            .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GGA_MASK as u16)
            .unwrap_or(0)
            | mask
                .gll
                .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16)
                .unwrap_or(0)
            | mask
                .gsa
                .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GSA_MASK as u16)
                .unwrap_or(0)
            | mask
                .gsv
                .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GSV_MASK as u16)
                .unwrap_or(0)
            | mask
                .rmc
                .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_RMC_MASK as u16)
                .unwrap_or(0)
    }
}
```

Instead, this happened:

<details>
  <summary>Clippy output</summary>

```
warning: failed to automatically apply fixes suggested by rustc to crate `nrf_modem`

after fixes were automatically applied the compiler reported errors within these files:

  * src\gnss.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: expected identifier, found keyword `if`
   --> src\gnss.rs:220:15
    |
220 |             | if mask
    |               ^^ expected identifier, found keyword

error: expected one of `,`, `:`, or `@`, found `mask`
   --> src\gnss.rs:220:18
    |
220 |             | if mask
    |                 -^^^^ expected one of `,`, `:`, or `@`
    |                 |
    |                 help: missing `,`

error: expected `,`
   --> src\gnss.rs:221:35
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                  ---              ^^
    |                  |
    |                  while parsing the fields for this pattern

error: expected one of `,`, `:`, or `@`, found `.`
   --> src\gnss.rs:221:17
    |
220 |             | if mask
    |                      -
    |                      |
    |                      expected one of `,`, `:`, or `@`
    |                      help: missing `,`
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                 ^ unexpected token

error: expected identifier, found keyword `else`
   --> src\gnss.rs:221:75
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                                                                           ^^^^ expected identifier, found keyword
    |
help: escape `else` to use it as an identifier
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } r#else { 0 }
    |                                                                           ++

error: expected one of `,` or `:`, found keyword `else`
   --> src\gnss.rs:221:75
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                                                                          -^^^^ expected one of `,` or `:`
    |                                                                          |
    |                                                                          help: missing `,`

error[E0422]: cannot find struct, variant or union type `gll` in this scope
   --> src\gnss.rs:221:18
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                  ^^^ not found in this scope

error[E0422]: cannot find struct, variant or union type `r#else` in this scope
   --> src\gnss.rs:221:75
    |
221 |                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |                                                                           ^^^^ not found in this scope

error[E0308]: mismatched types
   --> src\gnss.rs:219:23
    |
219 |         if mask.gga { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GGA_MASK as u16 } else { 0 }
    |         --------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-------------
    |         |             |
    |         |             expected `()`, found `u16`
    |         expected this to be `()`
    |
help: you might have meant to return this value
    |
219 |         if mask.gga { return nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GGA_MASK as u16; } else { 0 }
    |                       ++++++                                                 +

error[E0308]: mismatched types
   --> src\gnss.rs:219:81
    |
219 |         if mask.gga { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GGA_MASK as u16 } else { 0 }
    |         ------------------------------------------------------------------------^--
    |         |                                                                       |
    |         |                                                                       expected `()`, found integer
    |         expected this to be `()`
    |
help: you might have meant to return this value
    |
219 |         if mask.gga { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GGA_MASK as u16 } else { return 0; }
    |                                                                                 ++++++  +

error: aborting due to 10 previous errors

Some errors have detailed explanations: E0308, E0422.
For more information about an error, try `rustc --explain E0308`.
```
</details>

This is one of the diagnostics:

```
warning: use of `.then_some(..).unwrap_or(..)` can be written more clearly with `if .. else ..`
   --> src\gnss.rs:222:15
    |
222 |               | mask
    |  _______________^
223 | |                 .gll
224 | |                 .then_some(nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16)
225 | |                 .unwrap_or(0)
    | |_____________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#obfuscated_if_else
help: try
    |
222 ~             | if mask
223 +                 .gll { nrfxlib_sys::NRF_MODEM_GNSS_NMEA_GLL_MASK as u16 } else { 0 }
    |
```

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-pc-windows-msvc
release: 1.70.0
LLVM version: 16.0.2
```


### Additional Labels

_No response_

--- a/clippy_lints/src/methods/obfuscated_if_else.rs
+++ b/clippy_lints/src/methods/obfuscated_if_else.rs
@@ -1,6 +1,7 @@
 use super::OBFUSCATED_IF_ELSE;
 use clippy_utils::diagnostics::span_lint_and_sugg;
 use clippy_utils::eager_or_lazy::switch_to_eager_eval;
+use clippy_utils::get_parent_expr;
 use clippy_utils::source::snippet_with_applicability;
 use clippy_utils::sugg::Sugg;
 use rustc_errors::Applicability;
@@ -40,6 +41,17 @@ pub(super) fn check<'tcx>(
             if_then,
             snippet_with_applicability(cx, unwrap_arg.span, "..", &mut applicability)
         );
+
+        // To be parsed as an expression, the `if { … } else { … }` as the left operand of a binary operator
+        // requires parentheses.
+        let sugg = if let Some(parent_expr) = get_parent_expr(cx, expr)
+            && let ExprKind::Binary(_, left, _) = parent_expr.kind
+            && left.hir_id == expr.hir_id
+        {
+            format!("({sugg})")
+        } else {
+            sugg
+        };
 
         span_lint_and_sugg(
             cx,

--- a/tests/ui/obfuscated_if_else.fixed
+++ b/tests/ui/obfuscated_if_else.fixed
@@ -22,3 +22,29 @@ fn main() {
     if true { () } else { a += 2 };
     //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 }
+
+fn issue11141() {
+    // Parentheses are required around the left side of a binary expression
+    let _ = (if true { 40 } else { 17 }) | 2;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are required only for the leftmost expression
+    let _ = (if true { 30 } else { 17 }) | if true { 2 } else { 3 } | if true { 10 } else { 1 };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required around the right side of a binary expression
+    let _ = 2 | if true { 40 } else { 17 };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a cast
+    let _ = if true { 42 } else { 17 } as u8;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a deref
+    let _ = *if true { &42 } else { &17 };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a deref followed by a cast
+    let _ = *if true { &42 } else { &17 } as u8;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+}

--- a/tests/ui/obfuscated_if_else.fixed
+++ b/tests/ui/obfuscated_if_else.fixed
@@ -3,16 +3,22 @@
 
 fn main() {
     if true { "a" } else { "b" };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     if true { "a" } else { "b" };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 
     let a = 1;
     if a == 1 { "a" } else { "b" };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     if a == 1 { "a" } else { "b" };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 
     let partial = (a == 1).then_some("a");
     partial.unwrap_or("b"); // not lint
 
     let mut a = 0;
     if true { a += 1 } else { () };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     if true { () } else { a += 2 };
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 }

--- a/tests/ui/obfuscated_if_else.rs
+++ b/tests/ui/obfuscated_if_else.rs
@@ -3,16 +3,22 @@
 
 fn main() {
     true.then_some("a").unwrap_or("b");
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     true.then(|| "a").unwrap_or("b");
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 
     let a = 1;
     (a == 1).then_some("a").unwrap_or("b");
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     (a == 1).then(|| "a").unwrap_or("b");
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 
     let partial = (a == 1).then_some("a");
     partial.unwrap_or("b"); // not lint
 
     let mut a = 0;
     true.then_some(a += 1).unwrap_or(());
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
     true.then_some(()).unwrap_or(a += 2);
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 }

--- a/tests/ui/obfuscated_if_else.rs
+++ b/tests/ui/obfuscated_if_else.rs
@@ -22,3 +22,29 @@ fn main() {
     true.then_some(()).unwrap_or(a += 2);
     //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
 }
+
+fn issue11141() {
+    // Parentheses are required around the left side of a binary expression
+    let _ = true.then_some(40).unwrap_or(17) | 2;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are required only for the leftmost expression
+    let _ = true.then_some(30).unwrap_or(17) | true.then_some(2).unwrap_or(3) | true.then_some(10).unwrap_or(1);
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required around the right side of a binary expression
+    let _ = 2 | true.then_some(40).unwrap_or(17);
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a cast
+    let _ = true.then_some(42).unwrap_or(17) as u8;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a deref
+    let _ = *true.then_some(&42).unwrap_or(&17);
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+
+    // Parentheses are not required for a deref followed by a cast
+    let _ = *true.then_some(&42).unwrap_or(&17) as u8;
+    //~^ ERROR: this method chain can be written more clearly with `if .. else ..`
+}

--- a/tests/ui/obfuscated_if_else.stderr
+++ b/tests/ui/obfuscated_if_else.stderr
@@ -8,31 +8,31 @@ LL |     true.then_some("a").unwrap_or("b");
    = help: to override `-D warnings` add `#[allow(clippy::obfuscated_if_else)]`
 
 error: this method chain can be written more clearly with `if .. else ..`
-  --> tests/ui/obfuscated_if_else.rs:6:5
+  --> tests/ui/obfuscated_if_else.rs:7:5
    |
 LL |     true.then(|| "a").unwrap_or("b");
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { "a" } else { "b" }`
 
 error: this method chain can be written more clearly with `if .. else ..`
-  --> tests/ui/obfuscated_if_else.rs:9:5
+  --> tests/ui/obfuscated_if_else.rs:11:5
    |
 LL |     (a == 1).then_some("a").unwrap_or("b");
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if a == 1 { "a" } else { "b" }`
 
 error: this method chain can be written more clearly with `if .. else ..`
-  --> tests/ui/obfuscated_if_else.rs:10:5
+  --> tests/ui/obfuscated_if_else.rs:13:5
    |
 LL |     (a == 1).then(|| "a").unwrap_or("b");
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if a == 1 { "a" } else { "b" }`
 
 error: this method chain can be written more clearly with `if .. else ..`
-  --> tests/ui/obfuscated_if_else.rs:16:5
+  --> tests/ui/obfuscated_if_else.rs:20:5
    |
 LL |     true.then_some(a += 1).unwrap_or(());
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { a += 1 } else { () }`
 
 error: this method chain can be written more clearly with `if .. else ..`
-  --> tests/ui/obfuscated_if_else.rs:17:5
+  --> tests/ui/obfuscated_if_else.rs:22:5
    |
 LL |     true.then_some(()).unwrap_or(a += 2);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { () } else { a += 2 }`

--- a/tests/ui/obfuscated_if_else.stderr
+++ b/tests/ui/obfuscated_if_else.stderr
@@ -37,5 +37,53 @@ error: this method chain can be written more clearly with `if .. else ..`
 LL |     true.then_some(()).unwrap_or(a += 2);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { () } else { a += 2 }`
 
-error: aborting due to 6 previous errors
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:28:13
+   |
+LL |     let _ = true.then_some(40).unwrap_or(17) | 2;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(if true { 40 } else { 17 })`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:32:13
+   |
+LL |     let _ = true.then_some(30).unwrap_or(17) | true.then_some(2).unwrap_or(3) | true.then_some(10).unwrap_or(1);
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(if true { 30 } else { 17 })`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:32:48
+   |
+LL |     let _ = true.then_some(30).unwrap_or(17) | true.then_some(2).unwrap_or(3) | true.then_some(10).unwrap_or(1);
+   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { 2 } else { 3 }`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:32:81
+   |
+LL |     let _ = true.then_some(30).unwrap_or(17) | true.then_some(2).unwrap_or(3) | true.then_some(10).unwrap_or(1);
+   |                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { 10 } else { 1 }`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:36:17
+   |
+LL |     let _ = 2 | true.then_some(40).unwrap_or(17);
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { 40 } else { 17 }`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:40:13
+   |
+LL |     let _ = true.then_some(42).unwrap_or(17) as u8;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { 42 } else { 17 }`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:44:14
+   |
+LL |     let _ = *true.then_some(&42).unwrap_or(&17);
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { &42 } else { &17 }`
+
+error: this method chain can be written more clearly with `if .. else ..`
+  --> tests/ui/obfuscated_if_else.rs:48:14
+   |
+LL |     let _ = *true.then_some(&42).unwrap_or(&17) as u8;
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `if true { &42 } else { &17 }`
+
+error: aborting due to 14 previous errors
 

